odoo/addons/base/models/res_users.py

```diff
@@ -1127,6 +1144,27 @@ def _remove_reified_groups(self, values):
             else:
                 values1[key] = val
 
+        portal = self.env.ref('base.group_portal',raise_if_not_found=False)
+        public = self.env.ref('base.group_public',raise_if_not_found=False)
+        if add:
+            #If we choose Portal or Public as user type, only this group is added.
+            if portal and portal.id in add:
+                add = [portal.id]
+            elif public and public.id in add:
+                add = [public.id]
+
+        #If there are an existing user with two types, and we change another group.
+        #We leave only internal, to avoid raise an exception.
+        for user in self:
+            userTypeGrp = user.groups_id.filtered(lambda x : x.category_id.xml_id == 'base.module_category_user_type')
```

Comment:
We usually don't use `xml_id` directly, but rather: `x.category_id == self.env.ref('base.module_category_user_type'....)`

Double check it works, though.